server/src/db/filesystem_db.rs

```diff
@@ -168,11 +168,53 @@ impl DbTrait for FileSystemDb {
     }
 
     // PUT
+    fn update_page(&self, user_id: &Id<User>, page: Page) -> DbResult<Page> {
+        self.update_entity(user_id, page)
+    }
+
+    fn update_widget(&self, user_id: &Id<User>, widget: Widget) -> DbResult<Widget> {
+        self.update_entity(user_id, widget)
+    }
+
     fn update_bookmark(&self, user_id: &Id<User>, bookmark: Bookmark) -> DbResult<Bookmark> {
         self.update_entity(user_id, bookmark)
     }
 
     // DELETE
+    fn delete_page(&self, user_id: &Id<User>, page_id: &Id<Page>) -> DbResult {
+        let mut widgets = self.get_widgets(user_id)?;
+        widgets.retain(|b| &b.page_id == page_id);
+        for w in widgets {
+            self.delete_widget(user_id, &w.id)?;
+        }
+        let page_path = self.get_path(user_id, Some(page_id));
+
+        match std::fs::remove_file(page_path) {
+            Ok(_) => Ok(()),
+            Err(e) => match e.kind() {
+                std::io::ErrorKind::NotFound => Err(DbError::NotFound),
+                _ => Err(DbError::WhoopsieDoopsie),
+            },
+        }
+    }
+
+    fn delete_widget(&self, user_id: &Id<User>, widget_id: &Id<Widget>) -> DbResult {
+        let mut bookmarks = self.get_bookmarks(user_id)?;
+        bookmarks.retain(|b| &b.widget_id == widget_id);
+        for b in bookmarks {
+            self.delete_bookmark(user_id, &b.id)?;
+        }
+        let widget_path = self.get_path(user_id, Some(widget_id));
+
+        match std::fs::remove_file(widget_path) {
+            Ok(_) => Ok(()),
+            Err(e) => match e.kind() {
+                std::io::ErrorKind::NotFound => Err(DbError::NotFound),
+                _ => Err(DbError::WhoopsieDoopsie),
+            },
+        }
```

Comment:
This part is pretty much identical between bookmark, widget and page. Try to extract it into a function that's *generic* over the trait `DbEntity`. There are several of such generic functions already, take `store_entity` as an example.